### PR TITLE
Schema extracted to separate module

### PR DESCRIPTION
**Pull request by marcinwadon**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -306,7 +327,7 @@ lazy val dagShared = (project in file("modules/dag-shared"))
 lazy val dagL1 = (project in file("modules/dag-l1"))
   .enablePlugins(AshScriptPlugin)
   .enablePlugins(JavaAppPackaging)
-  .dependsOn(kernel, shared % "compile->compile;test->test", dagShared, sdk, testShared % Test)
+  .dependsOn(dagShared % "compile->compile;test->test", sdk, kernel, testShared % Test)
```
Comment (TheMMaciek) [inline build.sbt:330]: What about transitive dependencies? For example I'm almost sure that dagL1 depends on schema as well.

File: modules/core/src/main/scala/org/tessellation/http/routes/DagRoutes.scala [hunk 2]
```diff
@@ -7,8 +7,9 @@ import cats.syntax.functor._
 import org.tessellation.dag.domain.block.DAGBlock
 import org.tessellation.domain.cell.{L0Cell, L0CellInput}
 import org.tessellation.domain.dag.DAGService
-import org.tessellation.ext.http4s.{AddressVar, SnapshotOrdinalVar}
-import org.tessellation.security.signature.Signed
+import org.tessellation.ext.http4s.SnapshotOrdinalVar
```
Comment (TheMMaciek) [inline modules/core/src/main/scala/org/tessellation/http/routes/DagRoutes.scala:10]: Shouldn't `SnapshotOrdinalVar` also go to `schema` module?

File: modules/core/src/main/scala/org/tessellation/package.scala [hunk 3]
```diff
@@ -6,18 +6,18 @@ import cats.effect.Async
...
 
 import eu.timepit.refined.auto._
-import eu.timepit.refined.numeric.Interval
+import eu.timepit.refined.numeric.Greater
 import io.estatico.newtype.ops._
 
 package object tessellation {
 
-  type CoreKryoRegistrationIdRange = Interval.Closed[700, 799]
+  type CoreKryoRegistrationIdRange = Greater[100]
```
Comment (TheMMaciek) [inline modules/core/src/main/scala/org/tessellation/package.scala:20]: We don't want any constraint for the max registration number?

Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/package.scala:20]: we already discussed that and this is what @kpudlik did in his pull request.
if we're going to rediscuss it, let's create a separate task and handle/change all kryo id ranges

File: modules/core/src/main/scala/org/tessellation/http/routes/DagRoutes.scala (hunk 2 above)
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/http/routes/DagRoutes.scala:10]: probably yes, I'll check

File: build.sbt (hunk 1 above)
Comment (marcinwadon) [reply build.sbt:330]: in the current form, you were not able to handle all the dependencies and we used partially explicitly declared dependencies and partially implicitly (transitive). it's better to stay with one solution, so it would be way cleaner, to specify minimum set of libraries explicitly in the base package, and whatever is needed more on higher packages, should be declared separately. we cannot avoid transitive dependencies anyway, so let's use it instead of redeclaring them (and then noone is keeping an order there)

File: modules/kernel/src/main/scala/org/tessellation/kernel/package.scala [hunk 4]
```diff
@@ -13,7 +13,6 @@ package object kernel {
   type KernelKryoRegistrationId = KryoRegistrationId[KernelKryoRegistrationIdRange]
 
   val kernelKryoRegistrar: Map[Class[_], KernelKryoRegistrationId] = Map(
-    classOf[ProcessSnapshot] -> 400,
-    classOf[StateChannelSnapshotGist] -> 401
```
Comment (TheMMaciek) [inline modules/kernel/src/main/scala/org/tessellation/kernel/package.scala:17]: Shouldn't we somehow mark that the 401 was already used, to avoid the situation where we reuse that registration id for another class? Nothing probably will happen in this particular case, but I wonder if it shouldn't be the approach in general to register new classes under new (never used) ids only.

File: modules/schema/src/main/scala/org/tessellation/schema/transaction.scala [hunk 5]
```diff
@@ -67,7 +67,7 @@ object transaction {
     val _Ordinal: Lens[TransactionReference, TransactionOrdinal] = GenLens[TransactionReference](_.ordinal)
 
     def of[F[_]: Async: KryoSerializer](signedTransaction: Signed[Transaction]): F[TransactionReference] =
-      signedTransaction.value.hashF.map(TransactionReference(signedTransaction.ordinal, _))
+      signedTransaction.value.hashF.map(a => TransactionReference(signedTransaction.ordinal, a))
```
Comment (TheMMaciek) [inline modules/schema/src/main/scala/org/tessellation/schema/transaction.scala:70]: If `_` doesn't work anymore maybe call the value `h` or `hash` instead of `a`

File: modules/core/src/main/scala/org/tessellation/package.scala (hunk 3 above)
Comment (tbekas) [reply modules/core/src/main/scala/org/tessellation/package.scala:20]: Why `Greater[100]`? It will likely conflict with twitter chill registrations.

File: modules/schema/src/main/scala/org/tessellation/schema/transaction.scala (hunk 5 above)
Comment (marcinwadon) [reply modules/schema/src/main/scala/org/tessellation/schema/transaction.scala:70]: ah, good catch. it's a leftover

Comment (marcinwadon) [reply modules/schema/src/main/scala/org/tessellation/schema/transaction.scala:70]: fixed :+1:

File: modules/core/src/main/scala/org/tessellation/http/routes/DagRoutes.scala (hunk 2 above)
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/http/routes/DagRoutes.scala:10]: done :+1:

File: modules/shared/src/main/scala/org/tessellation/ext/kryo.scala [hunk 6]
```diff
@@ -38,6 +39,14 @@ object kryo {
 
   type KryoRegistrationId[C] = Int Refined C
 
+  implicit class MapRegistrationGreaterId[A](mapa: Map[Class[_], KryoRegistrationId[Greater[A]]]) {
+
+    def union[B](mapb: Map[Class[_], KryoRegistrationId[Greater[A]]]) =
```
Comment (TheMMaciek) [inline modules/shared/src/main/scala/org/tessellation/ext/kryo.scala:44]: The `B` here is not used

Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/kryo.scala]: fixed :+1:

File: modules/core/src/main/scala/org/tessellation/package.scala (hunk 3 above)
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/package.scala]: after discussion, we've decided to keep it `Greater[300]`, because twitter/chill registers more than 100 classes

File: modules/kernel/src/main/scala/org/tessellation/kernel/package.scala (hunk 4 above)
Comment (kpudlik) [reply modules/kernel/src/main/scala/org/tessellation/kernel/package.scala:17]: I think we should distinguish `never used` from `never serialized`. If it hasn't been serialized with these ids before then we should be fine. If you want to mark anyway then we can do some refined rule like `Not[400]` or sth.

File: build.sbt (hunk 1 above)
Comment (TheMMaciek) [reply build.sbt:330]: I just remember that there was a discussion and we wanted to avoid transitive dependencies hence the comment. Given that external libraries pull in their versions of dependencies with them there is a chance that by removing some explicitly declared library from the base module, the version of this library in the depending module may change (because it's neither declared explicitly in the base module anymore but still can be available through base module dependencies and was never declared explicitly in the depending module). If we drop by a major version I guess it can mean trouble. Also in such a situation we are losing the compiler warning about possible incompatibilities afaik. Anyway I'm not sure what's the best approach here, just recalling the discussion we had some time ago and giving my 2 cents.

File: modules/kernel/src/main/scala/org/tessellation/kernel/package.scala (hunk 4 above)
Comment (TheMMaciek) [reply modules/kernel/src/main/scala/org/tessellation/kernel/package.scala:17]: Just pointing to a possible problem. It may also not be that easy at some point to determine if a class was ever serialized or not.

File: build.sbt (hunk 1 above)
Comment (kpudlik) [reply build.sbt:330]: We can run `sbt dependencyBrowseGraph` or `sbt dependencyTree` to find transitive dependencies eventually.

Comment (marcinwadon) [reply build.sbt:330]: @TheMMaciek I didn't remove dependencies from base modules (shared, schema, etc). I just removed redeclaration (probably not all, I'd keep it as a continuation process), so:
- if shared requires kryo, then schema don't need to redeclare kryo and can use it, because depends on shared.
- if kernel requires droste, but not shared, then kernel should declare droste dependency
- if shared depend on external library (let's say http4s) that is dependant on fs2, we should anyway declare fs2 explicitly.

so tl;dr - no transitive dependencies from external libraries. only between our internal modules

File: modules/shared/src/main/scala/org/tessellation/ext/kryo.scala [hunk 7]
```diff
@@ -38,6 +39,14 @@ object kryo {
 
   type KryoRegistrationId[C] = Int Refined C
 
+  implicit class MapRegistrationGreaterEqualId[A](mapa: Map[Class[_], KryoRegistrationId[GreaterEqual[A]]]) {
```
Comment (tbekas) [inline modules/shared/src/main/scala/org/tessellation/ext/kryo.scala:42]: Why is it necessary? It seems like it's already covered by the existing `union`

File: build.sbt [hunk 8]
```diff
@@ -89,155 +74,191 @@ lazy val wallet = (project in file("modules/wallet"))
...
+      Libraries.http4sCirce,
+      Libraries.http4sJwtAuth,
+      Libraries.weaverCats,
+      Libraries.weaverDiscipline,
+      Libraries.weaverScalaCheck
     )
   )
 
-lazy val shared = (project in file("modules/shared"))
+lazy val schema = (project in file("modules/schema"))
```
Comment (tbekas) [inline build.sbt:150]: We should have an ADR describing what's the purpose of each module so that we don't mix them up when adding stuff (like we do currently).

**Review by TheMMaciek — APPROVED** [review]

**Review by kpudlik — APPROVED** [review]

**Review by tbekas — CHANGES_REQUESTED** [review]
What is the purpose/responsibility of the `schema` module? It's impossible to say if the correct classes were extracted without the context.

File: modules/shared/src/main/scala/org/tessellation/ext/kryo.scala (hunk 7 above)
Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/kryo.scala:42]: otherwise, it throws an error that something cannot be `Greater[300] Or Greater[300]`

**Comment by marcinwadon** [comment]
Closing the PR. I'll create new ones according to the ADR 12. (#526)
